src/darwin/Framework/CHIP/MTRDeviceType.mm

```diff
@@ -17,27 +17,42 @@
 
 #import <Matter/MTRDeviceType.h>
 
+#import "MTRDefines_Internal.h"
 #import "MTRDeviceTypeMetadata.h"
 #import "MTRLogging_Internal.h"
 
+#include <lib/support/CodeUtils.h>
 #include <lib/support/SafeInt.h>
 
 using namespace chip;
 
-@implementation MTRDeviceType
+MTR_DIRECT_MEMBERS
+@implementation MTRDeviceType {
+    const MTRDeviceTypeData * _meta;
+}
 
-- (nullable instancetype)initWithDeviceTypeID:(NSNumber *)id name:(NSString *)name isUtility:(BOOL)isUtility
+- (instancetype)initWithDeviceTypeData:(const MTRDeviceTypeData *)metaData
 {
-    if (!(self = [super init])) {
-        return nil;
-    }
-
-    _id = id;
-    _name = name;
-    _isUtility = isUtility;
+    self = [super init];
+    _meta = metaData;
     return self;
 }
 
+- (NSNumber *)id
+{
+    return @(_meta->id);
+}
+
+- (NSString *)name
+{
+    return const_cast<NSString *>(_meta->name);
```

Comment:
Why is const_cast OK here?  Needs documentation at the very least.

I guess it's OK because NSString is immutable, so in practice there is no difference between `NSString *` and `const NSString *`?

Reply:
Actually I think the cast isn't even needed, it's a `const` pointer, not a `const Nsstring`